### ISSUE TITLE
Fix expansion of scons variables in LIBS

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#2413]
This issue was originally created at: 2009-04-30 15:55:48.
This issue was reported by: `bradnelson`.
>bradnelson said at 2009-04-30 15:55:48

`SCons/Scanner/Prog.py` ===
```python
59a60,65
     # FROM THIS:
     # TO THIS:
     if env.get('LIBS_DO_SUBST'):
         libs = map(str, env.Flatten(env.subst_list(libs)))
     # END CHANGES.
```

>gregnoel said at 2009-05-13 21:16:37

Bug party triage.  SK to follow up with OP.

>stevenknight said at 2009-11-10 18:00:19

stevenknight => issues@scons

**Comment by mwichmann** [comment]
Eh?  Is this asking for a new construction var?  It's so incomplete I recommend just tossing.

**Comment by bdbaddog** [comment]
I'd argue there's no reason to not subst LIBS list?

**Comment by mwichmann** [comment]
I'm still trying to figure out what the request is (besides "12 years old", which is good for Whisky, and bad for bug reports).

**Comment by bdbaddog** [comment]
I think it's this:

Normally you'd have `LIBS=['a','b','c']`

This change is to support
```
LIB_SET_ONE=['a','b']
LIB_SET_TWO=['c']

LIBS='$LIB_SET_ONE $LIB_SET_TWO'
```

**Comment by mwichmann** [comment]
And that doesn't work now?  Does the change require adding/checking the env var? If so, a tiny manpage add would be needed as well.

**Comment by mwichmann** [comment]
So, is it correct to assume this was to go at the end of `_subst_libs`? There's not enough context in the included diff to be completely sure, but `SCons/Scanner/Prog.py` hasn't changed much.

**Comment by mwichmann** [comment]
So should this just always be done, or does it need a value to control it - from the snip:
```py
if env.get('LIBS_DO_SUBST')
```

**Comment by mwichmann** [comment]
> I'd argue there's no reason to not subst LIBS list?

Okay, some prospecting.  `LIBS` are subst'd in current code. That was added in a series of fixes a few years after this report by Alexey Klimkin with patch summaries like "Fix incomplete LIBS flattening and substitution in Program scanner".  Checking out the code as it was in 2009, substitutions were not done.  Ancient code:
```py
    try:
        libs = env['LIBS']
    except KeyError:
        # There are no LIBS in this environment, so just return a null list:
        return []
    if SCons.Util.is_String(libs):
        libs = string.split(libs)
    else:
        libs = SCons.Util.flatten(libs)
    # << this was the point of the proposed addition
```
Current code calls a helper:
```py
    try:
        libs = env['LIBS']
    except KeyError:
        # There are no LIBS in this environment, so just return a null list:
        return []
    libs = _subst_libs(env, libs)
```
and that function does substitution, directly if a string value; called recursively in case `libs` was a list.

I vote we close this as completed.

**Comment by mwichmann** [comment]
Annotating this: the fix seems to have been for SCons 2.3.5 as part of b22e9980df40305bb8beac3ea03bb045e01a031c

Closing.
